BUG: upgrade to PyQt6 to fix arm64 docker build

Alvaro-Kothe opened the pull request:
- [x] closes #61037 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

---
- I also refactored the qt imports to simplify the version selection
- The main goal of this PR is fix the ARM64 docker build. You may test it by running `docker build --platform linux/arm64 -t pandas-dev .` if you are on an `x86_64` system, but be warned that the build takes very long. 
- I added a new CI job to test the docker build on ARM64.
- Also, I have no clue how would I type annotate the return type of the `_import_module` auxiliary function.

Alvaro-Kothe committed:
fix: use `PyQt6` instead of `PyQt5`

Alvaro-Kothe committed:
chore: use latest pyqt version available on conda repo

Alvaro-Kothe committed:
ci(deps): install PyQt6 with pip

Alvaro-Kothe committed:
fix: replace pyqt with pyqt6

Alvaro-Kothe committed:
chore: use correct capitalization for PyQt

Alvaro-Kothe committed:
docs(whatsnew): add PyQt6 support to whats new

Alvaro-Kothe committed:
docs(install): keep old pyqt5 minimum version

Alvaro-Kothe committed:
fix: turn generator into tuple for correct error message

Alvaro-Kothe committed:
fix: let `getattr` raise `AttributeError` directly

Alvaro-Kothe committed:
Merge branch 'main' into fix/arm64-docker-build

Alvaro-Kothe committed:
Merge remote-tracking branch 'main' into fix/arm64-docker-build

Alvaro-Kothe committed:
doc: remove extra space between bar and PyQt6

Alvaro-Kothe commented:
Hi @mroeschke and @rhshadrach, can any of you review this PR when you have the time?

rhshadrach commented on doc/source/getting_started/install.rst:
```diff
@@ -340,12 +340,12 @@ Clipboard
+======================================================================================================================================== ================== =============== ==============
+Dependency                                                                                                                               Minimum Version    pip extra       Notes
+======================================================================================================================================== ================== =============== ==============
+`PyQt4 <https://pypi.org/project/PyQt4/>`__/`PyQt5 <https://pypi.org/project/PyQt5/>`__/`PyQt6 <https://pypi.org/project/PyQt6/>`__      5.15.9             clipboard       Clipboard I/O
```
You're listing the minimum version as 5.15.9 while changing the minimum-version action to PyQt6. In addition, it seems like PyQt4 should no longer be listed?

Alvaro-Kothe replied:
This change just extends the support to `PyQt6`, while keeping `PyQt4` and `PyQt5` compatibility intact.

> it seems like PyQt4 should no longer be listed?

It still supports `PyQt4`, so I think it should remain, but if you want me to de-list it I will do it.

> You're listing the minimum version as 5.15.9 while changing the minimum-version action to PyQt6

The goal of this change was just to replace the `PyQt5` entry in `requirements-dev.txt`, the changes in the CI dependencies were unintentional, but required to test the `PyQt6` changes in CI.

Pandas remains compatible with `PyQt4` and `PyQt5`. The problem is that it is not being tested in CI anymore. Maybe we can list `PyQt4`, `PyQt5` and `PyQt6` in the CI dependencies and parametrize the `qt` clipboard tests to test the different `PyQt` versions?

Alvaro-Kothe replied:
Thanks for the suggestion!

I was avoiding creating a big discrepancy between the CI jobs and ended up installing `qtpy`, `PyQt5` and `PyQt6` in all environments and parametrized the tests to use each of the Qt bindings. The problem with this approach is that the `qapp` fixture from `pytest-qt` doesn't handle the clipboard properly when the test is run with `PyQt5`, because `qapp` uses `PyQt6` to create the clipboard, and, for some reason, it can't read what's written with `PyQt5`.

Anyway, the parametrization approach works but creates complexity. The advantage of this is that every CI job tests `PyQt6` and `PyQt5`.

Splitting the CI is simpler and doesn't complicate the clipboard tests. I will leave it up to you if:

1. Test `PyQt5` in `actions-311-minimum_versions.yaml` and test `PyQt6` in the rest.
2. Leave the clipboard tests parametrized.

rhshadrach reviewed (commented):
It seems like you're dropping PyQt5 support here, is that right?

Alvaro-Kothe committed:
Merge remote-tracking branch 'upstream/main' into fix/arm64-docker-build

Alvaro-Kothe committed:
docs(install): remove `PyQt4` dependency

Alvaro-Kothe committed:
test(clipboard): parametrize qt tests to use multiple versions

Alvaro-Kothe committed:
test: add xfail mark when a qt version isn't available

Alvaro-Kothe committed:
ci: re-add PyQt5 as a CI dependency

Alvaro-Kothe committed:
ci(deps): rename PyQt5 to pyqt

Alvaro-Kothe committed:
test: be more specific about QtWidgets import

When PyQt6 is uninstalled, it is still importable, but `PyQt6.QtWidgets`
is no longer available.

Alvaro-Kothe commented:
I was able to reproduce the CI error on my machine, and confirmed that it happens when `PyQt6` and `PyQt5` are installed simultaneously in the same environment, which makes some `PyQt5` tests fail. When I uninstall `PyQt6` all `PyQt5` tests pass.

For the pre-commit error, the `validate-min-versions-in-sync` hook is trying to apply this patch in several CI dependencies:
```diff
-  - pyqt>=5.15.9
+  - pyqt>=6.7.1
```
Which is unintended, because there isn't `pyqt>=6` in conda-forge.

Alvaro-Kothe committed:
test: fix `ModuleNotFoundError`

Alvaro-Kothe committed:
test: match clipboard with PyQt

Alvaro-Kothe committed:
fix: fix PyQt5 dependency generator for conda

Alvaro-Kothe committed:
ci: move dependencies and re-add pyqt5 to min_versions

Alvaro-Kothe committed:
docs(install): make clipboard compatibility more explicit

Alvaro-Kothe committed:
Merge remote-tracking branch 'upstream/main' into fix/arm64-docker-build

Alvaro-Kothe committed:
Merge branch 'main' into fix/arm64-docker-build

Alvaro-Kothe committed:
test: undo changes in clipboard tests

Alvaro-Kothe committed:
chore: use `PyQt5` for minimum instead of `PyQt6`

The CI for `311-minimum_versions` tests the clipboard with `PyQt5`.

Alvaro-Kothe committed:
ci(clipboard): use `PyQt6` instead of `PyQt5` for non minimum tests

Alvaro-Kothe committed:
fix: remove pyqt5 from dev dependencies in favour of pyqt6

Alvaro-Kothe committed:
ci(docker-build): build on arm64

Alvaro-Kothe committed:
fix: add pyside6 support

Pyside2 support wasn't added because `pytest-qt>=4.5` doesn't support
it.

Reference:
<https://github.com/pytest-dev/pytest-qt/blob/master/CHANGELOG.rst#450-2025-07-01>

Alvaro-Kothe committed:
docs(clipboard): update note on clipboard dependencies

Set explicit system dependencies on Linux.

Make it explicit that Qt bindings support are limited on Linux and
states that it works best on Windows and MacOS.

The Qt support on Linux isn't reliable because it depends on an event
loop and user-input events to modify the clipboard. [1]

[1]: <https://doc.qt.io/qt-6/qclipboard.html#notes-for-x11-users>

Alvaro-Kothe commented on doc/source/user_guide/io.rst:
```diff
@@ -3928,7 +3928,28 @@ We can see that we got the same content back, which we had earlier written to th
+      sudo apt-get install xsel
+
+   Alternatively, you can install one of these Python packages,
+   but their support is limited on Linux and work more reliably
```
The Qt [clipboard documentation](https://doc.qt.io/qt-6/qclipboard.html#notes-for-x11-users) states that:

> the X11 clipboard is event driven, i.e. the clipboard will not function properly if the event loop is not running. Similarly, it is recommended that the contents of the clipboard are stored or retrieved in direct response to user-input events, e.g. mouse button or key presses and releases. You should not store or retrieve the clipboard contents in response to timer or non-user-input events.

The way that pandas currently uses the clipboard on Linux doesn't trigger any event, which doesn't trigger any clipboard modification.

Since I am on Linux (Wayland), here is an example of how it is not reliable:

```python
import sys
from PyQt6.QtWidgets import QApplication
from PyQt6.QtCore import QTimer, QEventLoop

app = QApplication(sys.argv)

app.clipboard().setText("Hello World!")
# CI tests this
print("clipboard before:", app.clipboard().text())
# clipboard before: Hello World!

# creates event loop and process events
loop = QEventLoop()
QTimer.singleShot(100, loop.quit)
loop.exec()

print("clipboard after :", app.clipboard().text())
# clipboard after :
```

Although the Qt bindings aren't reliable, the subprocess calling `wl-copy` works fine.

mroeschke reviewed (changes requested):
Thanks for working on this so far, but I am hesitant on supporting pyqt6 just for the Dockerfile, which unfortunately hasn't seen consistent maintenance over the years.
Additionally, our clipboard support was vendored from pyperclip, which I suppose didn't have much consideration for pyqt6 (maybe it didn't exist yet?)

If we were to make any improvements here, I would like to see pandas instead migrate to a different, non-vendored dependency for clipboard support instead, xref https://github.com/pandas-dev/pandas/issues/39834, or just deprecate support for this "IO" method if it's causing installation issues. I don't think we receive many bug reports or feature requests for this method either.

Alvaro-Kothe commented:
Thanks for the review.

> but I am hesitant on supporting pyqt6 just for the Dockerfile

I completely agree with you. Just to add pyqt6 support, I had to introduce several dependency changes in CI which I didn't mean to.

> I would like to see pandas instead migrate to a different, non-vendored dependency for clipboard support instead, xref https://github.com/pandas-dev/pandas/issues/39834, or just deprecate support for this "IO" method if it's causing installation issues.

Both approaches seem valid, I would be a little hesitant in using [spyoungtech/pyclip](https://github.com/spyoungtech/pyclip) because its last update was 2 years ago.

I am particularly fond of the deprecation approach, since it's not difficult to use `pyclip` or `pyperclip` to copy the data generated from pandas, for example:

```python
import pandas as pd
import io
import pyclip

df = pd.DataFrame(dict(a = [2, 5], b=[3, 4]))

with io.StringIO() as buf:
    df.to_csv(buf, sep="\t")
    pyclip.copy(buf.getvalue())
```

```console
$ python t.py

$ wl-paste
        a       b
0       2       3
1       5       4
```

Thanks for the review again! Closing.
